06_puppy/pfannerj/storer_test.go

```diff
@@ -0,0 +1,178 @@
+package main
+
+import (
+	"testing"
+
+	tassert "github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/suite"
+)
+
+type storerImpl int
+
+const (
+	syncImpl storerImpl = iota
+	mapImpl
+)
+
+const firstPuppyID uint32 = 1
+
+var (
+	firstPuppy = func() Puppy {
+		return Puppy{
+			Breed:  "Retriever",
+			Colour: "Golden",
+			Value:  9999.99,
+		}
+	}
+	modifiedPuppy = func() Puppy {
+		return Puppy{
+			Breed:  "Bitsa",
+			Colour: "Mixed",
+			Value:  1.99,
+		}
+	}
+	anotherPuppy = func() Puppy {
+		return Puppy{
+			Breed:  "Labrador",
+			Colour: "Black",
+			Value:  999.99,
+		}
+	}
+)
+
+type storerSuite struct {
+	suite.Suite
+	store Storer
+	impl  storerImpl
+}
+
+func (s *storerSuite) SetupTest() {
+	var err error
+	switch s.impl {
+	case syncImpl:
+		// create a new syncstore and add the first puppy
+		s.store = NewSyncStore()
+		puppySyncImpl := firstPuppy()
+		_, err = s.store.CreatePuppy(&puppySyncImpl)
+	case mapImpl:
+		// create a new mapstore and add the first puppy
+		s.store = NewMapStore()
+		puppyMapImpl := firstPuppy()
+		_, err = s.store.CreatePuppy(&puppyMapImpl)
+	default:
+		panic("Unrecognised storer implementation")
+	}
+	if err != nil {
+		panic("Failed to setup puppy test")
+	}
```

Comment:
Should be probably: 
```go
unc (s *storerSuite) SetupTest() {
	switch s.impl {
	case syncImpl:
		s.store = NewSyncStore()
	case mapImpl:
		s.store = NewMapStore()
	default:
		panic("Unrecognised storer implementation")
	}
	puppySyncImpl := firstPuppy()
	if _, err = s.store.CreatePuppy(&puppySyncImpl); err != nil {
		panic("Failed to setup puppy test")
	}
} 
better yet if `firstPuppy` returned already the pointer (or you made everything pointer free).

but let's take a step back.

IMO you are doubling up information unnecessarily with the TypeEnum (I know it's been suggested before, probably even used in the sample rest project and I think I might have been the first one at ANZ to use this (anti-?) pattern in the RefData project, not knowing about type switches / being generally a bit confused.)

Here is what I had in mind:
```go
type storerSuite struct {
	suite.Suite
	store puppy.Storer
}
 
func (s *storerSuite) SetupTest() {
        switch s.store.(type) {
	case store.MemStore:
		s.store = store.NewMemStore()
        // ...
	}
}

func TestStorer(t *testing.T) {
	suite.Run(t, &storerSuite{store: store.NewMemStore()}) // &storerSuite{store: &store.MemStore{}})
        // ....
}

```

alternatively you could use a factory type function: 

 ```go
type storerSuite struct {
	suite.Suite
	newStorer func() Storer
        store Storer
}

func (s *storerSuite) SetupTest() {
         s.store = s.newStorer()
}

func TestStorer(t *testing.T) {
	suite.Run(t, &storerSuite{newStorer: NewMemStore})
	suite.Run(t, &storerSuite{newStorer: NewSyncStore})
}
```


```

Reply:
I have changed over to this structure for the storer tests..  Just have one question that I will follow up with you when I get a chance...